node/src/Pos/Binary/Communication.hs

```diff
@@ -40,37 +40,55 @@ deriveSimpleBi ''MsgGetBlocks [
     ]]
 
 instance (HasCoreConstants, SscHelpersClass ssc) => Bi (MsgHeaders ssc) where
-  encode (MsgHeaders b) = encode b
-  decode = MsgHeaders <$> decode
+    encode = \case
+        (MsgHeaders b) -> encodeListLen 2 <> encode (0 :: Word8) <> encode b
+        (MsgNoHeaders t) -> encodeListLen 2 <> encode (1 :: Word8) <> encode t
+    decode = do
+        enforceSize "MsgHeaders" 2
+        tag <- decode @Word8
+        case tag of
+            0 -> MsgHeaders <$> decode
+            1 -> MsgNoHeaders <$> decode
+            t -> fail $ "MsgHeaders wrong tag: " <> show t
 
 instance (HasCoreConstants, SscHelpersClass ssc) => Bi (MsgBlock ssc) where
-  encode (MsgBlock b) = encode b
-  decode = MsgBlock <$> decode
+    encode = \case
+        (MsgBlock b) -> encodeListLen 2 <> encode (0 :: Word8) <> encode b
+        (MsgNoBlock t) -> encodeListLen 2 <> encode (1 :: Word8) <> encode t
+    decode = do
+        enforceSize "MsgBlock" 2
+        tag <- decode @Word8
+        case tag of
+            0 -> MsgBlock <$> decode
+            1 -> MsgNoBlock <$> decode
+            t -> fail $ "MsgBlock wrong tag: " <> show t
 
 -- deriveSimpleBi is not happy with constructors without arguments
 -- "fake" deriving as per `MempoolMsg`.
 -- TODO: Shall we encode this as `CBOR` TkNull?
 instance Bi MsgSubscribe where
-  encode MsgSubscribe = encode (42 :: Word8)
-  decode = do
-    x <- decode @Word8
-    when (x /= 42) $ fail "wrong byte"
-    pure MsgSubscribe
+    encode MsgSubscribe = encode (42 :: Word8)
+    decode = do
+        x <- decode @Word8
+        when (x /= 42) $ fail "wrong byte"
+        pure MsgSubscribe
 
 ----------------------------------------------------------------------------
 -- Protocol version info and related
 ----------------------------------------------------------------------------
 
 instance Bi HandlerSpec where
-  encode input = case input of
-    ConvHandler mname        -> encodeListLen 2 <> encode (0 :: Word8) <> encodeKnownCborDataItem mname
-    UnknownHandler word8 bs  -> encodeListLen 2 <> encode word8 <> encodeUnknownCborDataItem bs
-  decode = do
-    enforceSize "HandlerSpec" 2
-    tag <- decode @Word8
-    case tag of
-      0 -> ConvHandler        <$> decodeKnownCborDataItem
-      _ -> UnknownHandler tag <$> decodeUnknownCborDataItem
+    encode input = case input of
+        ConvHandler mname        ->
+            encodeListLen 2 <> encode (0 :: Word8) <> encodeKnownCborDataItem mname
+        UnknownHandler word8 bs  ->
+            encodeListLen 2 <> encode word8 <> encodeUnknownCborDataItem bs
+    decode = do
+        enforceSize "HandlerSpec" 2
+        tag <- decode @Word8
+        case tag of
+            0 -> ConvHandler        <$> decodeKnownCborDataItem
+            _ -> UnknownHandler tag <$> decodeUnknownCborDataItem
```

Comment:
It wasn't introduced in this change, but `decode` is not left-inverse to `encode` here: an encoded unknown handler with `0 :: Word8` will deserialize to conv handler.

Reply:
How bad is it? Can we fix it in a separate PR?